NET-337: broker: storage node metrics have null values for Cassandra related fields

bearni95 opened the pull request:
**Important changes**
- `VolumeLogger.close` is now an async method, fixing disconnection await
- Added `this.client.isConnected` check before publishing to temporarily address the scenario when the client has been closed but there is still a pending-to-fulfill promise that pushed metrics.
___

Earlier the metrics data was assumed to be in format:
```
{
  "writeBytes": 123,  // seconds
  "readyBytes": 456, // seconds
}
```

Now fixed to read the correct format, which is:
```
{
  "writeBytes": {
    "rate": 123000. // milliseconds
  },
  "readyBytes": {
    "rate": 456000 // milliseconds
  }
}
```

Assuming that data in readBytes and writeBytes is always of type `{rate: number, ...}` instead of raw numbers, as seen in:
https://github.com/streamr-dev/network-monorepo/blob/b3c47960669b2b22803f4ddd56467ec1ab6108ad/packages/network/src/helpers/MetricsContext.ts#L67


When Cassandra is disabled it's metrics will contain 0:
```
storage: {
  bytesWrittenPerSec: 0,
  bytesReadPerSec: 0,
}
```
May be worth considering to remove the `storage` entry altogether from the metrics.

The fix implemented in this PR has not automated tests. We updated `per-node-metrics.test.ts` to check the format of the report, but it doesn't test the actual data generation (lines 129-134 and 147-152) logic. It makes more sense to divide `per-node-metrics.test.ts` to separate tests that assert the actual data, and those tests could be inside `Storage.test.ts` etc.?

bearni95 committed:
fix: null cassandra values, improved tests

bearni95 committed:
fix: polished test details

timoxley commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -138,8 +142,10 @@ export class StreamMetrics {
                     if (metricsReport.metrics['broker/cassandra']) {
-                        this.report.storage.bytesWrittenPerSec = throttledAvg(this.report.storage.bytesWrittenPerSec, (metricsReport.metrics['broker/cassandra']) ? (metricsReport.metrics['broker/cassandra'].writeBytes as any).rate: 0)
-                        this.report.storage.bytesReadPerSec = throttledAvg(this.report.storage.bytesReadPerSec, (metricsReport.metrics['broker/cassandra']) ? (metricsReport.metrics['broker/cassandra'].readBytes as any).rate : 0)
+                        // @ts-expect-error
```
What is ts complaining about in these places? Can we fix it at the source?

timoxley replied:
I wonder if using `?.` would make ts happy, e.g. `metricsReport.metrics['broker/cassandra']?.writeBytes?.rate || 0`

bearni95 replied:
The logic ended up being a bit more involved but the @ts-expect-error is gone anyways

timoxley reviewed (approved):
needs conflict fix but otherwise lgtm.

bearni95 committed:
test: merging main into branch to resolve broken dependencies

bearni95 committed:
test: 03-08-2021 corrections applied

bearni95 committed:
test: updated to accept 0 values since some metrics have no feasable time to get measured during test time

harbu commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -124,8 +125,12 @@ export class StreamMetrics {
-                    this.report.storage.bytesReadPerSec = (metricsReport.metrics['broker/cassandra']) ? metricsReport.metrics['broker/cassandra'].readBytes : 0
+
+                    const cassandraWrittenBytesPerMs = metricsReport.metrics['broker/cassandra'].writtenBytes as {rate: number,total: number, last: number}
+                    this.report.storage.bytesWrittenPerSec = cassandraWrittenBytesPerMs.rate / 1000 || 0
```
Are we sure that the `rate` provided by the `MetricsContext` is in milliseconds? I recall it being in seconds with a window size of 5.

bearni95 replied:
After checking the measurements are provided in seconds, with the logic around the 5s window being handled before the metrics are populated

harbu commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -124,8 +125,12 @@ export class StreamMetrics {
+                    const cassandraWrittenBytesPerMs = metricsReport.metrics['broker/cassandra'].writtenBytes as {rate: number,total: number, last: number}
+                    this.report.storage.bytesWrittenPerSec = cassandraWrittenBytesPerMs.rate / 1000 || 0
+                    const cassandraReadBytesPerMs = metricsReport.metrics['broker/cassandra'].readBytes as {rate: number,total: number, last: number}
+                    this.report.storage.bytesReadPerSec = cassandraReadBytesPerMs.rate / 1000 || 0
```
Same as above.

harbu commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -137,10 +142,11 @@ export class StreamMetrics {
+                    const cassandraWrittenBytesPerMs = metricsReport.metrics['broker/cassandra'].writtenBytes as {rate: number,total: number, last: number}
+                    const cassandraReadBytesPerMs = metricsReport.metrics['broker/cassandra'].readBytes as {rate: number,total: number, last: number}
+
+                    this.report.storage.bytesWrittenPerSec = throttledAvg(this.report.storage.bytesWrittenPerSec, cassandraWrittenBytesPerMs.rate / 1000 || 0)
```
Same as above.

harbu commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -137,10 +142,11 @@ export class StreamMetrics {
+                    const cassandraReadBytesPerMs = metricsReport.metrics['broker/cassandra'].readBytes as {rate: number,total: number, last: number}
+
+                    this.report.storage.bytesWrittenPerSec = throttledAvg(this.report.storage.bytesWrittenPerSec, cassandraWrittenBytesPerMs.rate / 1000 || 0)
+                    this.report.storage.bytesReadPerSec = throttledAvg(this.report.storage.bytesReadPerSec, cassandraReadBytesPerMs.rate / 1000 || 0)
```
Same as above.

harbu commented on packages/broker/test/integration/per-node-metrics.test.ts:
```diff
@@ -3,10 +3,12 @@ import {startTracker, Tracker} from 'streamr-network'
 import { startBroker, createClient, STREAMR_DOCKER_DEV_HOST, createTestStream } from '../utils'
 import { Wallet } from 'ethers'
 import { Broker } from '../broker'
+import { Todo } from '../types'
```
Let's avoid using `Todo` in new code. It is there as a placeholder type for old code.

harbu commented on packages/broker/test/integration/per-node-metrics.test.ts:
```diff
@@ -48,9 +50,40 @@ const fillMetrics = async (client: StreamrClient, count: number, nodeAddress: st
+const waitForMessage = (
+    stream: string,
+    client: StreamrClient, 
+): Promise<Todo> => {
```
```suggestion
): Promise<any> => {
```

harbu commented on packages/broker/test/integration/per-node-metrics.test.ts:
```diff
@@ -48,9 +50,40 @@ const fillMetrics = async (client: StreamrClient, count: number, nodeAddress: st
+        try {
+            client.subscribe({stream}, (res: Todo) => {
+                resolve(res)
+            })
```
Unfortunately `any` is the type that streamr-client gives us for `subscribe` cb function parameter :shrug: 
```suggestion
            client.subscribe({stream}, (res: any) => {
                resolve(res)
            })
```

harbu commented on packages/broker/test/integration/per-node-metrics.test.ts:
```diff
@@ -48,9 +50,40 @@ const fillMetrics = async (client: StreamrClient, count: number, nodeAddress: st
+    })
+}
+
+const expectMetrics = (res: Todo) => {
```
```suggestion
const expectMetrics = (res: Record<string, unknown>) => {
```

harbu commented on packages/broker/test/integration/per-node-metrics.test.ts:
```diff
@@ -81,21 +115,13 @@ describe('per-node metrics', () => {
-            name: 'storageNode',
+            name: 'storageNode', 
             privateKey: storageNodeAccount.privateKey,
+            networkPort: networkPort,
```
AFAIK `networkPort` is no longer used, where did this come from?

bearni95 replied:
without it the storage node doesn't start and the tests crash on the `before` sections

bearni95 replied:
Actually Eric was right, without the network port it does work fine, removed it

bearni95 committed:
fix: legacy support for cassandra stream metrics

bearni95 committed:
fix: multiple corrections

- [x] move the cassandra expect errors into actually fluent code
- [x] remove the other redundant stuff that checksmultiple times
- [x] promisifySubscribe in tests; improve signature, better naming
- [x] add helper method for repeated lines
- [x] change /1000 divisions on ms
- [x] check if the first bit in the ifs for cassandra are required to operate
- [x] assign the numberic value for writtenBytes when found instead of 0
- [x] change interfaces to only include {rate: number}
- [x] validate structure rather than contents on expectMetrics
- [x] rename expectMetrics

bearni95 committed:
fix: style changes, replaced 'Todo' with proper typings in tests

bearni95 committed:
fix: cleanup, networkPort gone, need to look into tests

bearni95 committed:
fix: readings in seconds, improved tests

bearni95 committed:
fix: better naming for variables referencing seconds

bearni95 committed:
fix: changed back test trackerPort, unintentional

bearni95 committed:
fix: small touches

teogeb commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -137,9 +143,11 @@ export class StreamMetrics {
-                        this.report.storage.bytesWrittenPerSec = throttledAvg(this.report.storage.bytesWrittenPerSec, (metricsReport.metrics['broker/cassandra']) ? (metricsReport.metrics['broker/cassandra'].writeBytes as any).rate: 0)
-                        this.report.storage.bytesReadPerSec = throttledAvg(this.report.storage.bytesReadPerSec, (metricsReport.metrics['broker/cassandra']) ? (metricsReport.metrics['broker/cassandra'].readBytes as any).rate : 0)
+                    if (metricsReport.metrics['broker/cassandra']){
+                        const bytesWrittenPerSec = (metricsReport.metrics['broker/cassandra'].writeBytes as {rate: number}).rate ||  0
```
As you mentioned "Assuming that data in readBytes and writeBytes is always of type {rate: number, ...}", this fallback `0` is most likely redundant

teogeb commented on packages/broker/test/integration/per-node-metrics.test.ts:
```diff
@@ -81,21 +121,12 @@ describe('per-node metrics', () => {
+            name: 'storageNode', 
             privateKey: storageNodeAccount.privateKey,
             trackerPort,
+            wsPort,
```
The common setup is separate the storage nodes and publishe nodes. Maybe worth considering to use the same architecture?

teogeb reviewed (approved):
Good, let's add the tests in a later PR?

bearni95 committed:
fix: added check for client connected before reporting metrics

harbu commented on packages/broker/src/StreamMetrics.ts:
```diff
@@ -253,7 +261,7 @@ export class StreamMetrics {
 
     private async publishReport(): Promise<unknown> {
-        if (!this.stopped && this.targetStreamId) {
+        if (!this.stopped && this.client.isConnected && this.targetStreamId) {
```
What is this `this.client.isConnected` check for?

harbu commented on packages/broker/src/VolumeLogger.ts:
```diff
@@ -168,7 +168,7 @@ export class VolumeLogger {
 
         // Report metrics to Streamr stream
-        if (this.client instanceof StreamrClient && this.legacyStreamId !== undefined) {
+        if (this.client instanceof StreamrClient && this.client.isConnected && this.legacyStreamId !== undefined) {
```
What is this `this.client.isConnected` check for?

bearni95 committed:
fix: proper close sequence to avoid checking the client.isConnected under metrics

teogeb commented on packages/broker/src/VolumeLogger.ts:
```diff
@@ -290,7 +290,7 @@ export class VolumeLogger {
     }
 
-    close() {
+    async close() {
```
Some test runs show a warning in when I run  `per-node-metrics.test.ts`:
```
WARN [2021-08-12T12:17:36.013] (StreamMetrics       ): Failed to publish to stream 0x4F2693C21088D490774f9FBa7d135707894E2875/streamr/node/metrics/hour due to: Error: Needs connection but – connection: disconnected & wants: DISCONNECTED
autoConnect: false & autoDisconnect: false with 1 handles.
autoConnect disabled automatically after explicit call to connect/disconnect().
```

The warning is not show every time, but in about 50% of the runs.

Were these warnings the reason why `stop` + `close` handling was modified in the latest commit? 

If this error is not related to Cassandra `null` values, we could maybe open another PR to fix the behaviour.

teogeb replied:
Could the root cause be that there is a ongoing publish task when the Broker is closed?

E.g. would the following be a possible scenario:

1) A timer task triggers the report publish task. It calls `this.client.publish(this.targetStreamId, this.report)`. That method returns a promise,  so the task it is completed only once the promise returns. At any moment before that, we may need an open connection. 

2) Some time after publish a user or a test stops the Broker. The `broker.stop()` methods is called. Among other things it closes the `VolumeLogger`. There we  set `StreamMetrics` flag `stopped=true` and disconnect the client . Setting the `stopped=true` disables any future publishing, but it doesn't affect any ongoing publish tasks.

So if `broker.stop()` is called when `client.publish()` promise is pending, the `publish()` fail as `this.client.ensureDisconnected()` already disconnected the client.

bearni95 replied:
The check for `client.isConnected()` should deal with this scenario until the spin-off PR can fulfill this fix.

bearni95 committed:
fix: check for client.isConnected() before publishing metrics

teogeb reviewed (approved):
Maybe worth adding a note to the PR description about the changes that were good improvements but not related to this PR:

- VolumeLogger.close is now an async method, fixed disconnection await
- added this.client.isConnected check before publishing (and what this change fixes)

bearni95 committed:
fix: merge with main

bearni95 committed:
fix: post merge cleanup

bearni95 committed:
build: unmanageable merged leads to failing tests, keeping for reference before reverting to pre-merge commit

bearni95 committed:
revert: back to 5d27728d7bf0093d9be0e038d84e91cfda0784be, approved and pending of a merge from main

juslesan committed:
Merge branch 'main' into NET-337

bearni95 committed:
fix: merged Main into branch
